kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/batch/DynamicBatchWriter.java

```diff
@@ -0,0 +1,273 @@
+                                   String topic,
+                                   Set<Schema> schemas) throws BigQueryConnectException,
+                                                               InterruptedException {
+    int currentIndex = 0;
+    while (currentIndex < elements.size()) {
+      try {
+        // handle case where no splitting is necessary:
+        if (elements.size() <= currentBatchSize) {
+          writer.writeRows(table, elements, topic, schemas);
+          currentIndex = elements.size();
+          return;
+        }
+
+        int elementsLeft = elements.size() - currentIndex;
+        int numBatches = (int)Math.ceil(elementsLeft / (currentBatchSize * 1.0));
+        int minBatchSize = elementsLeft / numBatches;
+        int spareRows = elementsLeft % numBatches;
+
+        for (int batchCount = 0; batchCount < numBatches; batchCount++) {
+          // the first spareRows batches have an extra row in them.
+          int batchSize = batchCount < spareRows ? minBatchSize + 1 : minBatchSize;
+          int endIndex = Math.min(currentIndex + batchSize, elements.size());
+          writer.writeRows(table, elements.subList(currentIndex, endIndex), topic, schemas);
+          currentIndex = endIndex;
+        }
+      } catch (BigQueryException exception) {
+        if (isBatchSizeError(exception)) {
+          // immediately decrease batch size and try again with remaining elements.
+          logger.debug("Batch size error during establishedWriteAll, reducing batch size.");
+          decreaseBatchSize();
+          contSuccessCount = 0;
+        } else {
+          throw new BigQueryConnectException(
+              String.format("Failed to write to BigQuery table %s", table),
+              exception);
+        }
+      }
+    }
+
+    contSuccessCount++;
+    if (contSuccessCount >= CONT_SUCCESS_COUNT_BUMP) {
+      logger.debug("{} successful establishedWriteAlls in a row, increasing batchSize to {}",
+                   contSuccessCount, currentBatchSize);
+      contSuccessCount = 0;
+      increaseBatchSize();
+    }
+  }
+
+  /**
+   * @param exception the {@link BigQueryException} to check.
+   * @return true if this error is an error that can be fixed by retrying with a smaller batch
+   *         size, or false otherwise.
+   */
+  private boolean isBatchSizeError(BigQueryException exception) {
+    if (exception.code() == BAD_REQUEST_CODE
+        && exception.error() == null
+        && exception.reason() == null) {
+      // 400 with no error or reason represents a request that is more than 10MB. This is not
+      // documented but is referenced slightly under "Error codes" here:
+      // https://cloud.google.com/bigquery/quota-policy
+      // (by decreasing the batch size we can eventually expect to end up with a request under
+      // 10MB)
+      return true;
+    } else if (exception.code() == BAD_REQUEST_CODE && INVALID_REASON.equals(exception.reason())) {
```

Comment:
based on how it's documented I think it's the reason.  At least, quotaExceeded is a reason: https://github.com/wepay/kafka-connect-bigquery/blob/master/kcbq-connector/src/main/java/com/wepay/kafka/connect/bigquery/write/BigQueryWriter.java#L157